### PR TITLE
Enhancement/optimized loc provider

### PR DESCRIPTION
**Pull request by rdblue**

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/LocationProviders.java [hunk 1]
```diff
@@ -104,9 +106,10 @@ public String newDataLocation(String filename) {
   }
 
   static class ObjectStoreLocationProvider implements LocationProvider {
-    private static final Function<Object, Integer> HASH_FUNC =
-        Transforms.bucket(Integer.MAX_VALUE).bind(Types.StringType.get());
 
+    private static final HashFunction HASH_FUNC = Hashing.murmur3_32_fixed();
+    private static final BaseEncoding BASE64_ENCODER = BaseEncoding.base64Url().lowerCase().omitPadding();
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/LocationProviders.java:111]: Using the Guava encoder allows us to just omit padding. That way we don't have to worry about lengths and get the most bits in the final hash.

File: core/src/main/java/org/apache/iceberg/LocationProviders.java [hunk 2]
```diff
@@ -104,9 +106,10 @@ public String newDataLocation(String filename) {
...
+    private final ThreadLocal<byte[]> temp = ThreadLocal.withInitial(() -> new byte[4]);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/LocationProviders.java:112]: This reuses buffers so we avoid allocations.

File: core/src/main/java/org/apache/iceberg/LocationProviders.java [hunk 3]
```diff
@@ -167,5 +170,12 @@ private static String pathContext(String tableLocation) {
 
       return resolvedContext;
     }
+
+    private String computeHash(String fileName) {
+      byte[] bytes = temp.get();
+      HashCode hash = HASH_FUNC.hashString(fileName, StandardCharsets.UTF_8);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/LocationProviders.java:176]: Using `hashString` avoids converting the buffer to bytes and copying.

**Comment by singhpk234** [comment]
closing pr merged upstream !

**Pull request closed without merging**